### PR TITLE
cpu/stm32l1: fix issue when  writing NULL bytes to eeprom

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR fixes an issue where category 1,2,3 stm32l1xxx devices can't write NULL bytes to eeprom (reference manual pag 78). If this is attempted the app freezes.

### Testing procedure

You need a category 1,2 or 3 board. eg: im880b or lobaro-lorabox. Run:

`make -C tests/periph_eeprom BOARD=im880b test
`

without this PR failes.

### Issues/PRs references

Can use #11315 to test. ~~Depends on #11355~~.

### CODE REVIEW

**Comment by fjmolinas** [comment]
ping @aabadie

File: cpu/stm32_common/periph/eeprom.c [hunk 1]
```diff
@@ -36,20 +36,73 @@ extern void _wait_for_pending_operations(void);
...
+
+    /* Check the parameters */
+    assert(addr <= EEPROM_SIZE + EEPROM_START_ADDR);
+
+    /* Wait for last operation to be completed */
+    _wait_for_pending_operations();
+
+    /* stm32l1xxx cat 1, 2 &3 can't write NULL bytes RefManual p78*/
+    if(data != (uint8_t) 0x00)
+    {
```
Comment (aabadie) [inline cpu/stm32_common/periph/eeprom.c]: Still need a bit of unscrustify ;)

Comment (fjmolinas) [reply cpu/stm32_common/periph/eeprom.c]: yep, addressed!

**Comment by fjmolinas** [comment]
@aabadie comments addressed!

**Comment by kYc0o** [comment]
I just ran it on a lobaro-lorabox and works properly on master:

```console
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
facosta@meryl:~/git/RIOT-OS/RIOT$ BOARD=lobaro-lorabox make -C tests/periph_eeprom/ flash test
make: Entering directory '/home/facosta/git/RIOT-OS/RIOT/tests/periph_eeprom'
Building application "tests_periph_eeprom" for "lobaro-lorabox" with MCU "stm32l1".

"make" -C /home/facosta/git/RIOT-OS/RIOT/boards/lobaro-lorabox
"make" -C /home/facosta/git/RIOT-OS/RIOT/core
"make" -C /home/facosta/git/RIOT-OS/RIOT/cpu/stm32l1
"make" -C /home/facosta/git/RIOT-OS/RIOT/cpu/cortexm_common
"make" -C /home/facosta/git/RIOT-OS/RIOT/cpu/cortexm_common/periph
"make" -C /home/facosta/git/RIOT-OS/RIOT/cpu/stm32_common
"make" -C /home/facosta/git/RIOT-OS/RIOT/cpu/stm32_common/periph
"make" -C /home/facosta/git/RIOT-OS/RIOT/cpu/stm32l1/periph
"make" -C /home/facosta/git/RIOT-OS/RIOT/drivers
"make" -C /home/facosta/git/RIOT-OS/RIOT/drivers/periph_common
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/auto_init
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/isrpipe
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/newlib_syscalls_default
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/pm_layered
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/shell
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/shell/commands
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/stdio_uart
"make" -C /home/facosta/git/RIOT-OS/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  13300	    140	   2676	  16116	   3ef4	/home/facosta/git/RIOT-OS/RIOT/tests/periph_eeprom/bin/lobaro-lorabox/tests_periph_eeprom.elf
/home/facosta/git/RIOT-OS/RIOT/dist/tools/stm32loader/stm32loader.py -p /dev/ttyUSB0 -e -u -S -l 0x1ff -w /home/facosta/git/RIOT-OS/RIOT/tests/periph_eeprom/bin/lobaro-lorabox/tests_periph_eeprom.bin
Bootloader version 0x31
Chip id 0x429, STM32L1
Writing binary
test
/home/facosta/git/RIOT-OS/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyUSB0" -b "115200" --set-rts 0
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2019-04-18 15:11:59,948 - INFO # Connect to serial port /dev/ttyUSB0
Welcome to pyterm!
Type '/exit' to exit.
2019-04-18 15:12:00,955 - INFO # �main(): This is RIOT! (Version: 2019.04-devel-930-g98ed4)
2019-04-18 15:12:00,956 - INFO # EEPROM read write test
2019-04-18 15:12:00,956 - INFO # 
2019-04-18 15:12:00,958 - INFO # Please refer to the README.md for more details
2019-04-18 15:12:00,958 - INFO # 
2019-04-18 15:12:00,959 - INFO # EEPROM start addr:	0x08080000
2019-04-18 15:12:00,960 - INFO # EEPROM size:		4096
> 2019-04-18 15:12:02,892 - INFO #  test
2019-04-18 15:12:03,060 - INFO # SUCCESS

make: Leaving directory '/home/facosta/git/RIOT-OS/RIOT/tests/periph_eeprom'
```

**Comment by fjmolinas** [comment]
Thanks for testing @kYc0o!!

**Comment by fjmolinas** [comment]
@aabadie Think you could give it a test too?

**Comment by aabadie** [comment]
I tested this PR on im880b (stm32l151cb), nz32-sc151 (stm32l151rc) and nucleo-l152re (stm32l152re). The 2 last ones works on master and with this PR. The first one, doesn't work on master.

stm32l151cb and stm32l151rc are cat 1 and 2. Are you sure the bug is also on cat 2 ? @kYc0o reported that it works on lobaro (stm32l151cb) which is also cat1. This is weird.

I'm fine merging this PR (which is cut out from #10026 but I'd like to have a good reason for applying this change (and that some comments could be written in the code to explain it)

**Comment by fjmolinas** [comment]
@aabadie well... it is indeed weird I had actually only tested on `im880b` and since the ReferenceManual says that you can't write a NULL byte I though it would be true for cat 1, 2, 3 ... but lobaro and im880b use the same MCU so I'm actually a little confused about this... I'll have to do some digging into erratas... arrgg

**Comment by fjmolinas** [comment]
I took another look at the RM and there is contradicting statements:

In table Table 20. p.77:

> 7. The Fast Byte Write, Fast Half Word Write, Byte Write and Half Word Write can be used only to write a NON NULL byte/half
> word.

In table Table 21. p.79:

> 7. In contrast to Cat.1 devices, in Cat.2, Cat.3, Cat.4, Cat.5 and Cat.6 devices the Fast Byte Write, Fast Half Word Write, Byte
> Write and Half Word Write can used to write a NULL byte/half word.

I'll update the PR to guard only for Cat.1 devices.

**Comment by fjmolinas** [comment]
@aabadie @kYc0o found why it works for lobaro, this is a problem that affect only cat1 devices. And the lobaro is cat2 since it is a `stm32l151cb-a` cpu and not a `stm32l151cb`.

I know guarded only for `stm32l151cb`.

File: cpu/stm32_common/periph/eeprom.c [hunk 2]
```diff
@@ -36,6 +37,61 @@ extern void _wait_for_pending_operations(void);
...
+    _wait_for_pending_operations();
+
+    *(__IO uint32_t *)addr = data;
+}
+#endif
+
+static void _write_byte(uint32_t addr, uint8_t data)
+{
+    /* Check the parameters */
+    assert(addr <= EEPROM_SIZE + EEPROM_START_ADDR);
```
Comment (aabadie) [inline cpu/stm32_common/periph/eeprom.c]: This seems unrelated and means the assert is checked for each byte written. I would drop this.

File: cpu/stm32_common/periph/eeprom.c [hunk 3]
```diff
@@ -45,7 +101,7 @@ size_t eeprom_read(uint32_t pos, uint8_t *data, size_t len)
     DEBUG("Reading data from EEPROM at pos %" PRIu32 ": ", pos);
     for (size_t i = 0; i < len; i++) {
         _wait_for_pending_operations();
-        *p++ = *(uint8_t *)(EEPROM_START_ADDR + pos++);
+        *p++ = *(__IO uint8_t *)(EEPROM_START_ADDR + pos++);
```
Comment (aabadie) [inline cpu/stm32_common/periph/eeprom.c]: This should go in its own commit as it's a bit unrelated. It also requires test on other supported families...

**Comment by aabadie** [comment]
And go ;)
